qiskit/passmanager/base_optimization_tasks.py

```diff
@@ -0,0 +1,231 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Baseclasses for the Qiskit passmanager optimization tasks."""
+from __future__ import annotations
+
+import logging
+import time
+from abc import abstractmethod, ABC
+from collections.abc import Iterable, Callable, Iterator, Sequence
+from typing import Any
+
+from .exceptions import PassManagerError
+from .propertyset import PassState, PropertySet, FencedPropertySet, RunState
+
+logger = logging.getLogger(__name__)
+
+
+class OptimizerTask(ABC):
+    """A definition of optimization task.
+
+    The optimizer task takes a passmanager IR, and outputs new passmanager IR
+    after program optimization. Optimization task can rely on the :class:`.PassState`
+    to communicate intermediate state with other tasks.
+    """
+
+    _state: PassState | None
+
+    @property
+    def state(self) -> PassState:
+        """A local state information associated with this optimization workflow."""
+        if self._state is None:
+            # Allocate state information as needed basis.
+            # This slightly reduces memory footprint for instantiation.
+            self._state = PassState()
+        return self._state
+
+    @property
+    def property_set(self) -> PropertySet:
+        """Property set of this flow controller."""
+        return self.state.property_set
+
+    @property_set.setter
+    def property_set(self, new_property_set: PropertySet):
+        self.state.property_set = new_property_set
+
+    @abstractmethod
+    def execute(
+        self,
+        passmanager_ir: Any,
+        state: PassState | None = None,
+        callback: Callable = None,
+    ) -> Any:
+        """Execute optimization task for input passmanager IR.
+
+        Args:
+            passmanager_ir: Passmanager IR to optimize.
+            state: A local state information associated with this optimization workflow.
+            callback: A callback function which is caller per execution of optimization task.
+
+        Returns:
+            Optimized passmanager IR.
+        """
+        pass
```

Comment:
I was assuming the same IR because LLVM was in mind. I agree staged pass manager with flexibility of stage-specific IR allows more sophisticated optimization without ugly hack. This is more like MLIR though. I'm not really familiar with its concept, but as far as I understand, we need to introduce IR (dialect) conversion pass, and each dialect should implement operation (which is python method in our case), and the conversion pass is basically a mapper of the operations. Because we don't need to distinguish the conversion pass from analysis/transform passes, allowing different IR doesn't hurt current design, i.e. conversion pass can return different IR. 

The pass is agnostic to dialect, so task should work as long as the IR supports required methods -- likely duck typing? So I just need to relax the restriction of the IR type. Probably I need better handling of calibration for DAG transform pass though.

Let me know if I'm missing something important.

Reply:
This all sounds sensible to me.  

> The pass is agnostic to dialect, so task should work as long as the IR supports required methods -- likely duck typing?

What sorts of methods do we think a pass manager would ever need to call on the IR itself?  I'd have thought that a `Task` / converter would be the only thing that actually needed to know anything about the IR other than its type, but maybe I'm missing something you've got in mind?

---

Just for greater visibility for others reading this / posterity on a decision right now: We might need/want multiple IRs whereas one of LLVM's design strengths is that it _doesn't_ have multiple during the middle-end (but ofc frontend and backend components have different representations) because Qiskit's compilation pipelines certainly at the moment don't have a front/middle/backend split. For lowering to quantum devices, it doesn't seem immediately obvious that the same split in classical compilers will hold.

When we're talking about circuit lowering, the closest thing to a frontend representation we have is probably OQ3, but for most purposes, I we can skip over discussion of that component, because it doesn't feature in our optimisation / lowering.  However, we don't really have much (if any) of a middle-end / backend split; if anything, Qiskit does the bit that feels more analogous to classical backends _first_ in layout+routing+basis-translation, and only after that does it do quantum optimisation ("middle" end).  We're chopping and churning around that, especially with the high-level synthesis work going on, and more backend/optimisation-aware synthesis early on, so it's not super clear how things will end up being best.

I think for now, we'll probably want to design our pass manager under the assumption that there might be multiple IRs.  If nothing else, it will allow us a backwards-compatible way to upgrade the compiler IR incrementally, without breaking user plugins written for old versions of the IR (such as a sort of `DAGCircuitV2` that I'm hoping to dig into next year).

Reply:
I agree with you. Pass manager and controller doesn't need to know about IR. Only passes rely on the IR's method. I updated the documentation in [e20949a](https://github.com/Qiskit/qiskit/pull/10127/commits/e20949af2a7cce3066b611b148d4b4749c2f61a7) so that future developer can introduce multiple IRs without breaking the original design.